### ISSUE TITLE
Avoid ! for negation

### ISSUE THREAD

**Issue by mokagio** [iflix-letsplay/swift-style-guide#8]
There's a lot to be said on the value of readable code (see [this great talk](https://www.youtube.com/watch?v=B3X_nhJ61Ak)).

The more readable the code is the easier it will be to understand, which will make it easier to review and change.

Last year [I was shown](https://youtu.be/iihrmC_h2Gk?t=18m51s) the dangers of using `!` for negations. The TL;DR is that that little single character is very sneaky, and easy to miss.

Since then I started avoiding `!` and always write conditionals in an explicit form:

```diff
- if !someCondition {
+ if someCondition == false {
  // do something
}
```

This takes longer to write, yes. But the code -imho- is much easier to understand.

**Comment by TimKregerIflix** [comment]
Yeah agreed. I also have a problem with Ternary operators for a similar reason, I've seen a lot of bugs hidden by unclear ternary operations that would be more readble and traceabe by using a conditional statement

**Comment by mokagio** [comment]
Open an issue for the ternary operator @TimKregerIflix  😉 .

**Comment by nhisyam** [comment]
I am OK with this suggestion..

**Comment by martinpilch** [comment]
Does it apply also for this case?
`tableView.setEditing(!tableView.isEditing, animated: true)`
vs
`tableView.setEditing(tableView.isEditing == false, animated: true)`

**Comment by mokagio** [comment]
Good point @martinpilch... 🤔 

Tbh it looks quite weird to me used as a function parameter, but I guess the same concerns of `!` being easily missed when reading the code still apply 🤔 .

Because of that I'd say yes.

**Comment by mokagio** [comment]
Actually thinking about it... the example you provided is not really a negation is more a "change the bool value", right?

In such case maybe it's appropriate?

**Comment by mokagio** [comment]
Please cast your vote with a 👍 reaction to the option you'd prefer.

**Comment by mokagio** [comment]
Always avoid `!`.

### Preferred

```swift
if condition == false { doStuff() }

tableView.setEditing(tableView.isEditing == false, animated: true)
```

### Not Preferred

```swift
if !condition { doStuff() }

tableView.setEditing(!tableView.isEditing, animated: true)
```

**Comment by mokagio** [comment]
Avoid `!` only when used in conditionals.

### Preferred

```swift
if condition == false { doStuff() }

tableView.setEditing(!tableView.isEditing, animated: true)
```

### Not Preferred

```swift
if !condition { doStuff() }
```

**Comment by mokagio** [comment]
The use of `!` for negation should be up to the code author.

**Comment by iamariffikri** [comment]
sure, I agree. 👍

**Comment by mokagio** [comment]
Locked the conversation. This is the team's verdict on the matter

<img width="929" alt="screen shot 2017-10-06 at 3 08 06 pm" src="https://user-images.githubusercontent.com/1218433/31264195-29641822-aab3-11e7-9ea6-3bffd9e4b5f0.png">

**Comment by mokagio** [comment]
Closed by #21.
